Better error reporting

dlespiau opened the pull request:
We noticed a few bad things:

- Agent events were all classified under one error because events with the same stack frame are aggregated together.
- Bootstrap events didn't have any stack frame associated with them.
- raven's Exception object prefixes error messages with the type of the `error`. The issue is that most errors we used are directly from `errors.New()` or indirectly from `fmt.Errorf()`. They have the same type: `*errors.errorString`. Sentry picked that up as the error name, which lead to having a walll of "errors.errorString" errors instead of more interesting messages.
- We weren't reporting some of the errors in agent when trying to talk to Kubernetes.

leth commented on agent/main.go:
```diff
@@ -63,12 +64,15 @@ func setLogLevel(logLevel string) error {
+	formatted := fmt.Sprintf("%s: %s", msg, err)
+	log.Error(formatted)
+
+	tags := map[string]string{
```
we shoud set these at startup with `raven.SetTagsContext` :)

leth commented on pkg/sentry/sentry.go:
```diff
@@ -0,0 +1,30 @@
+// been sent to sentry.
+func Capture(message string, skipFrames int, tags map[string]string) {
+	// We want to at least skip the Capture() function itself.
+	if skipFrames < 1 {
```
Shouldn't this consistently `+= 1` to skipFrames? That way the caller only has to count the number of frames they want to skip.

leth commented on bootstrap/main.go:
```diff
@@ -132,14 +133,14 @@ func exitNoCapture(msg string, args ...interface{}) {
 }
 
-func capture(msg string, args ...interface{}) {
+func capture(skipFrames int, msg string, args ...interface{}) {
```
The args here are only slightly different to Capture; can we reorder them for consistency?

dlespiau replied:
The issue here is that msg, a format string, is usually just before the variadic parameters. we could put skip first everywhere but if feels a it odd.

leth replied:
Ah okay 👍

dlespiau commented:
Update PR:

- Rebased on top of master.
- Used SetTagContext for global flags.
- Renamed Capture to CaptureAndWait (same name as raven).
- Changed semantics of skipFrames to not include sentry.
- Change argument order to be able to write variadic wrappers that take the same argument order.
- Retested a bit as it's quite a few changes.

leth commented on pkg/sentry/sentry.go:
```diff
@@ -0,0 +1,31 @@
+// This call is synchronous in that it will not return until the message has
+// been sent to sentry.
+func CaptureAndWait(skipFrames uint, message string, tags map[string]string) {
+	// We want to skip the Capture() function itself.
```
Method name is out of date int the comment!

dlespiau replied:
Fixed! I think I liked Capture better after all :) I'll stop changing my mind and leave it as CaptureAndWait for now.
